File: apps/ios/Packages/DropCore/Sources/DropCore/InMemoryCommentsRepository.swift
import Foundation

/// 테스트와 프리뷰(`-dropPreview`)용 리포지토리. 네트워크 없이 같은 계약을 지킨다.
public final class InMemoryCommentsRepository: CommentsRepository, @unchecked Sendable {
    private let lock = NSLock()
    private var comments: [NoteComment]

    /// 실패 경로를 시험하기 위한 손잡이.
    public var loadError: Error?
    public var createError: Error?
    public var mutationError: Error?

    /// 로드를 원하는 시점까지 붙잡아 두기 위한 손잡이 — 겹친 로드를 재현한다.
    public var beforeLoad: (@Sendable () async -> Void)?

    public private(set) var loadCallCount = 0
    public private(set) var createCallCount = 0
    /// 다듬기(trim)가 실제로 걸렸는지 보기 위한 것.
    public private(set) var lastCreatedBody: String?

    public init(comments: [NoteComment] = []) {
        self.comments = comments
    }

    public func loadComments(noteID: String) async throws -> [NoteComment] {
        withLock { loadCallCount += 1 }
        await beforeLoad?()
        if let loadError { throw loadError }
        return withLock {
            comments.filter { $0.noteID == noteID }.sorted { $0.createdAt < $1.createdAt }
        }
    }

    public func loadCommentCounts() async throws -> [String: Int] {
        if let loadError { throw loadError }
        return withLock {
            comments.reduce(into: [:]) { counts, comment in
                counts[comment.noteID, default: 0] += 1
            }
        }
    }

    public func createComment(noteID: String, body: String) async throws -> NoteComment {
        withLock {
            createCallCount += 1
            lastCreatedBody = body
        }
        if let createError { throw createError }
        let comment = NoteComment(
            id: UUID().uuidString,
            noteID: noteID,
            body: body,
            createdAt: Date(),
            updatedAt: Date()
        )
        withLock { comments.append(comment) }
        return comment
    }

    public func deleteComment(id: String) async throws {
        if let mutationError { throw mutationError }
        withLock { comments.removeAll { $0.id == id } }
    }

    private func withLock<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }
}

File: apps/mobile/packages/drop_core/lib/src/in_memory_comments_repository.dart
import 'comments_repository.dart';
import 'note_comment.dart';

var _idCounter = 0;

String _nextId() {
  _idCounter += 1;
  return '메모리댓글-$_idCounter';
}

/// 테스트와 프리뷰용 리포지토리. 네트워크 없이 같은 계약을 지킨다.
/// DropCore `InMemoryCommentsRepository.swift` 대응.
class InMemoryCommentsRepository implements CommentsRepository {
  final List<NoteComment> _comments;

  /// 실패 경로를 시험하기 위한 손잡이.
  Object? loadError;
  Object? createError;
  Object? mutationError;

  /// 로드를 원하는 시점까지 붙잡아 두기 위한 손잡이 — 겹친 로드를 재현한다.
  Future<void> Function()? beforeLoad;

  int loadCallCount = 0;
  int createCallCount = 0;

  /// 다듬기(trim)가 실제로 걸렸는지 보기 위한 것.
  String? lastCreatedBody;

  InMemoryCommentsRepository({List<NoteComment> comments = const []})
      : _comments = [...comments];

  @override
  Future<List<NoteComment>> loadComments(String noteId) async {
    loadCallCount += 1;
    await beforeLoad?.call();
    final loadError = this.loadError;
    if (loadError != null) throw loadError;
    return _comments.where((comment) => comment.noteId == noteId).toList()
      ..sort((a, b) => a.createdAt.compareTo(b.createdAt));
  }

  @override
  Future<Map<String, int>> loadCommentCounts() async {
    final loadError = this.loadError;
    if (loadError != null) throw loadError;
    final counts = <String, int>{};
    for (final comment in _comments) {
      counts[comment.noteId] = (counts[comment.noteId] ?? 0) + 1;
    }
    return counts;
  }

  @override
  Future<NoteComment> createComment({
    required String noteId,
    required String body,
  }) async {
    createCallCount += 1;
    lastCreatedBody = body;
    final createError = this.createError;
    if (createError != null) throw createError;
    final comment = NoteComment(
      id: _nextId(),
      noteId: noteId,
      body: body,
      createdAt: DateTime.now().toUtc(),
      updatedAt: DateTime.now().toUtc(),
    );
    _comments.add(comment);
    return comment;
  }

  @override
  Future<void> deleteComment(String id) async {
    final mutationError = this.mutationError;
    if (mutationError != null) throw mutationError;
    _comments.removeWhere((comment) => comment.id == id);
  }
}
